### PR TITLE
Feature/battles api

### PR DESCRIPTION
**Pull request by k-szewczyk**
I've added API which allows you to create battles
Main features:
 - Only staff can POST data to /battles/ endpoint
 - After sending request serializers create() method creates Fight object, which handles battle system logic
 - Posted data is validated in cases if:
     - Selected heroes fought with each one previously (They can fight only once)
     - If races of selected heroes can fight
     - If both of selected heroes are alive

### CODE REVIEW

File: backend/battles/serializers.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        fight = Fight(validated_data['attendees'])
+        fight.battle_loop()
+        return fight.battle
+
+    def validate(self, attrs):
+        attendees = attrs['attendees']
+        annotated_attendees = Hero.objects.get_annotations().filter(Q(user=attendees[0]) | Q(user=attendees[1]))
+        for attendee in annotated_attendees:
+            if not attendee.is_alive:
+                raise ValidationError(f'Couldn\'t create fight player {attendee} is dead')
```
Comment (Rafal-Maszlej) [inline backend/battles/serializers.py]: I guess `hero` is dead, not `player`? :-P

File: backend/battles/serializers.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        races = [hero.race for hero in annotated_attendees]
+        if not races[0] in races[1].can_fight_with.all():
+            raise ValidationError(f'These two races can\'t fight with each other')
+
+        if len(Battle.objects.filter(attendees__user=attendees[0]).filter(attendees__user=attendees[1])) > 0:
```
Comment (Rafal-Maszlej) [inline backend/battles/serializers.py]: You shouldn't use `len` with a queryset

Comment (natkam) [reply backend/battles/serializers.py]: ...unless you want to evaluate it for some reason ;) Read about queryset evaluation in the [documentation](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#when-querysets-are-evaluated).

File: backend/halloffame/tests.py [hunk 3]
```diff
@@ -1,8 +1,8 @@
-from django.urls import reverse
 from rest_framework import test, status
+from rest_framework.reverse import reverse
 
+from halloffame import factories
 from halloffame.models import Hero
-from . import factories
 
 
 class HallOfFamePermissionsTestCase(test.APITestCase):
```
Comment (Rafal-Maszlej) [inline backend/halloffame/tests.py]: Duplicated class name ;-)

File: backend/battles/tests.py [hunk 4]
```diff
@@ -1,3 +1,102 @@
...
+    def test_one_hero_loses_battle(self):
+        fight = Fight([self.hero1, self.hero2])
+        fight.battle_loop()
+
+        queryset = Hero.objects.get_annotations()
+
+        won_counter = queryset.first().battles_won + queryset.last().battles_won
+        lost_counter = queryset.first().battles_lost + queryset.last().battles_lost
+
+        self.assertEqual(True, (won_counter == 1) & (lost_counter == 1))
```
Comment (Rafal-Maszlej) [inline backend/battles/tests.py]: `assertTrue`

**Pull request merged**